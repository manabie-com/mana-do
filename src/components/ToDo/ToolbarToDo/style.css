.Todo__toolbar {
  display: flex;
  justify-content: space-between;
  margin-top: 24px;
}

.Todo__toolbar .Action__checkAll {
  display: flex;
  align-items: center;
}

.Todo__tabs {
  display: flex;
  justify-content: center;
}

.Todo__tabs > *:not(:last-child) {
  margin-right: 20px;
}

.Todo__tabs .Action__btn {
  color: var(--primary-color);
  font-weight: 500;
  display: flex;
  align-items: center;
  padding: 10px 20px;
  border-radius: 8px;
}

.Todo__tabs .Action__btn-selected {
  background: var(--primary-color);
  color: white;
}
