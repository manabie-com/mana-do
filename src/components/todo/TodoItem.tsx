import React, { useEffect, useRef, useState } from 'react';
import { TodoStatus, Todo } from '../../models/todo';
import useClickOutside from '../../hooks/useClickOutside';

interface TodoItemProps {
  todo: Todo;
  onDelete?: (id: string) => void;
  onComplete?: (id: string, isComplete: boolean) => void;
  onUpdateContent?: (id: string, content: string) => void;
}

const TodoItem = (props: TodoItemProps) => {
  const [isUpdateContent, setIsUpdateContent] = useState<boolean>(false);
  const [editingContent, setEditingContent] = useState<string>();
  const itemRef = useRef<HTMLDivElement>(null);
  const editingContentRef = useRef<HTMLTextAreaElement>(null);

  useClickOutside(itemRef, () => {
    if (isUpdateContent === true) {
      setIsUpdateContent(false);
    }
  });

  useEffect(() => {
    if (isUpdateContent) {
      const length = editingContentRef.current?.value.length || null;
      editingContentRef.current?.setSelectionRange(length, length);
      editingContentRef.current?.focus();
    }
  }, [isUpdateContent]);

  const onEnableUpdateContent = (e: React.MouseEvent<HTMLParagraphElement>) => {
    // double click
    if (e.detail === 2) {
      setEditingContent(props.todo.content);
      setIsUpdateContent(true);
    }
  };

  const onUpdateTodoContent = (e: React.KeyboardEvent<HTMLTextAreaElement>) => {
    if (e.key === 'Enter' && editingContent?.trim()) {
      props.onUpdateContent &&
        props.onUpdateContent(props.todo.id, editingContent.trim());
      setIsUpdateContent(false);
    }
  };

  return (
    <div ref={itemRef} className="item">
      {isUpdateContent ? (
        <textarea
          ref={editingContentRef}
          className="item__content item__content-edit"
          value={editingContent}
          onChange={(e) => setEditingContent(e.target.value)}
          onKeyDown={onUpdateTodoContent}
        />
      ) : (
        <p
          className="item__content item__content-noedit"
          onClick={onEnableUpdateContent}
        >
          {props.todo.content}
        </p>
      )}

      <div className="item__actions">
        <button
          className="item__delete"
          onClick={() => props.onDelete && props.onDelete(props.todo.id)}
        >
          X
        </button>
        <div className="action__checkbox">
          <input
            id={`completed-${props.todo.id}`}
            type="checkbox"
            checked={TodoStatus.COMPLETED === props.todo.status}
            onChange={(e) =>
              props.onComplete &&
              props.onComplete(props.todo.id, e.target.checked)
            }
          />
          <label htmlFor={`completed-${props.todo.id}`}>Completed</label>
        </div>
      </div>
    </div>
  );
};

export default TodoItem;
