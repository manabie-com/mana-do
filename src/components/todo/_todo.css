
button {
  outline: none;
  border: none;
  box-shadow: 2px 0 2px currentColor;
  border-radius: 4px;
  min-height: 32px;
  min-width: 80px;
  padding: 4px 8px;
  color: steelblue;
  font-weight: 600;
}

button:hover {
  opacity: 0.85;
  border: 1px solid rgba(70,130,180, 0.7);
}

input[type="checkbox"] {
  width: 24px;
  height: 24px;
  box-shadow: none;
  border: none;
  outline: none;
  cursor: pointer;
  width: 1.5em;
  height: 1.5em;
}

input[type="checkbox"]:focus {
  box-shadow: none;
  border: none;
  outline: none;
}

input {
  min-height: 36px;
  border: none;
  outline: none;
  padding: 0 12px;
  box-shadow: 2px 0 4px rgba(0,0,0, 0.2);
  border-radius: 4px;
}

input:focus {
  box-shadow: 1px 0 9px rgba(0,0,0, 0.25);
}

.ToDo__Wrapper {
  background: url('./../../assets/images/bg_ocean.jpeg');
  background-repeat: no-repeat;
  background-size: cover;
  width: 100vw;
  height: 100vh;
  display: flex;
  justify-content: center;
  align-items: center;
}

.ToDo__container {
  border: 1px solid rgba(0,0,0, 0.4);
  border-radius: 8px;
  max-width: 700px;
  min-width: 350px;
  height: 60vh;
  width: 50vw;
  margin-top: 5rem;
  padding: 24px;
  box-shadow: 2px 2px 1px rgba(0,0,0, 0.09),
              3px 2px 3px rgba(0,0,0, 0.05);
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  background-color: rgba(70,130,180, 0.25);
  animation: form-appear 2s;
  opacity: 1;
}

@keyframes form-appear {
  from {
    opacity: 0;
    transform: translateY(-70px);
  }
  to {
    opacity: 1;
  }
}

.Todo__creation {
  display: flex;
  width: 100%;
}

input.TodoItem__input {
  width: 80%;
  background-color: rgba(255,255,255, 0.1);
  color: white;
  transition: 1s;
}

input.TodoItem__input:focus {
  box-shadow: 2px 0 4px rgba(213, 250, 1, 0.7);
}

input.TodoItem__input:hover {
  box-shadow: 2px 0 4px rgba(250, 113, 1, 0.7);
}

.Todo__input {
  flex: 1 1;
  background-color: rgba(70,130,180, 0.7);
  color: white;
  transition: 1s;
}

.Todo__input::placeholder {
  color: rgba(240, 247, 179, 0.7);
}

.Todo__input:focus {
  box-shadow: 2px 0 4px rgba(213, 250, 1, 0.7);
}

.Todo__input:hover {
  box-shadow: 2px 0 4px rgba(250, 113, 1, 0.7);
}

.ToDo__list {
  display: flex;
  flex-direction: column;
  margin-top: 1.5rem;
  width: 100%;
  height: 100%;
  overflow-y: auto;
  border-right: solid 1px rgba(127, 255, 212, 0.9);
  border-left: solid 1px rgba(127, 255, 212, 0.9);
  border-bottom: solid 2px rgba(127, 255, 212, 0.9);
  border-bottom-left-radius: 20px;
  border-bottom-right-radius: 20px;
  background-color: rgba(127, 255, 212, 0.2);
  box-shadow: 4px 0 4px currentColor;
}

.ToDo__item {
  display: flex;
  align-items: center;
  justify-content: space-between;
  margin: 4px 2px;
  padding: 2px 4px;
  border: 1px solid rgba(70,130,180, 0.4);
  border-radius: 15px;
  background-color: rgba(70,130,180, 0.4);
  color: white;
  animation: item-appear 0.4s;
  cursor: pointer;
  transition: 0.5s;
  font-size: 16px;
  min-height: 2.5em;
}

.ToDo__item:hover {
  background-color: rgba(240, 247, 181, 0.3);
}

@keyframes item-appear {
  from {
    opacity: 0;
    transform: translateY(30px);
  }
  to {
    opacity: 1;
  }
}

.ToDo__item > span {
  flex: 1 1;
  text-align: left;
  margin-left: 8px;
  margin-right: 8px;
  display: -webkit-box;
  -webkit-line-clamp: 2;
  word-break: break-word;
  -webkit-box-orient: vertical;
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 5px;
  font-size: 16px;
  line-height: 24px;
}

.Todo__content {
  flex: 1 1;
}

.Todo__delete {
  outline: none;
  border: none;
  width: 32px;
  height: 32px;
  min-width: auto;
  min-height: auto;
  box-shadow: none;
  border-radius: 50%;
  display: flex;
  justify-content: center;
  align-items: center;
  cursor: pointer;
  background-color: rgba(255, 0, 0, 0.6);
  color: white;
}

.Todo__action, .Todo__delete {
  width: 24px;
  height: 24px;
  flex-shrink: 0;
}

.Todo__toolbar {
  display: flex;
  justify-content: space-between;
  margin-top: 24px;
}

.Todo__tabs {
  display: flex;
  justify-content: center;
}

.Todo__tabs > *:not(:last-child) {
  /* margin-right: 8px; */
  border-right: 2px solid rgba(70,130,180, 0.4);
}

.Action__btn {
  cursor: pointer;
  background-color: aquamarine;
  transition-duration: 0.5s;
}

.Action__btn:hover {
  transform: translateY(-5px);
}

.Action__btn.active {
  background-color: rgba(70,130,180, 0.8);
  color: white;
}

.Action__btn:disabled {
  cursor: not-allowed;
}

.use-tooltip {
  position: relative;
  display: inline;
}
  
.use-tooltip:hover .tooltip {
  display: block;
  position: fixed;
  z-index: 3;
  left: 50%;
  top: -20px;
  transform: translateX(-50%);
  min-width: 80vw;
  text-align: left;
  display: -webkit-box;
  -webkit-line-clamp: 5;
  word-break: break-word;
  -webkit-box-orient: vertical;
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 0.5em;
  font-size: 16px;
  line-height: 24px;
  color: steelblue;
  font-weight: 500;
}
  
.tooltip {
  display: none;
  color: white;
  background-color: rgb(240, 247, 181, 0.7);
  border-radius: 20px;
  margin: 2em;
}

@media screen and (min-width: 1201px) {
}
@media screen and (max-width: 1200px) {
}
@media screen and (max-width: 1024px) {
  .ToDo__container {
    width: 70vw;
  }
}
@media screen and (max-width: 768px) {
  .ToDo__container {
    width: 100vw;
  }
}
@media screen and (max-width: 480px) {
}