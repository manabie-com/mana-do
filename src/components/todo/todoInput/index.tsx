import React from "react";
import { ITodoProps } from "./ITodoProps";
import "./todoInput.scss";

const TodoInput = (props: ITodoProps): JSX.Element => {
  return (
    <div className="todo-input">
      <input {...props} className="todo-input__input" />
    </div>
  );
};

export default TodoInput;
