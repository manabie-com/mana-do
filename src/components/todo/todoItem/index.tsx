import React, { useRef, useState } from "react";
import { Constants } from "../../../constants";
import { useTodo } from "../../../hooks/useTodo";
import {
  deleteTodo,
  updateTodoContent,
  updateTodoStatus,
} from "../../../store/actions";
import { isTodoCompleted } from "../../../utils";
import { ITodoItemProps } from "./ITodoItemProps";
import "./todoItem.scss";

const TodoItem = ({ todo }: ITodoItemProps): JSX.Element => {
  const { dispatch } = useTodo();

  const [editable, setEditable] = useState<boolean>(false);
  const [editContent, setEditContent] = useState<string>(todo.content);
  const inputRef = useRef<HTMLInputElement>(null);

  const onChangeContent = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.key === Constants.ENTER && editContent) {
      dispatch(updateTodoContent(todo.id, editContent));
      setEditContent("");
      setEditable(false);
    }
  };

  const onUpdateTodoStatus = (e: React.ChangeEvent<HTMLInputElement>) => {
    dispatch(updateTodoStatus(todo.id, e.target.checked));
  };

  const onDeleteTodo = () => {
    // we should show popup confirmation before clear all
    if (window.confirm("Are you sure to delete ?")) {
      dispatch(deleteTodo(todo.id));
    }
  };

  return (
    <div className="todo-item">
      <input
        type="checkbox"
        checked={isTodoCompleted(todo)}
        onChange={onUpdateTodoStatus}
        className="todo-item__checkbox"
        data-testid="checkbox"
      />
      <label
        onDoubleClick={() => {
          setEditable(true);
          setEditContent(todo.content);
          inputRef.current?.focus();
        }}
      >
        {editable ? (
          <input
            autoFocus
            value={editContent}
            onChange={(e) => setEditContent(e.target.value)}
            type="text"
            onBlur={() => {
              setEditable(false);
              setEditContent(todo.content);
            }}
            onKeyDown={onChangeContent}
            ref={inputRef}
            className="todo-item__edit"
          />
        ) : (
          <span>{todo.content}</span>
        )}
      </label>
      <a
        href="#/"
        data-testid="delete"
        className="todo-item__delete"
        onClick={onDeleteTodo}
      >
        &times;
      </a>
    </div>
  );
};

export default TodoItem;
