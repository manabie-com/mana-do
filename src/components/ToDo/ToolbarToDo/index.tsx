import React, { useState } from "react";
import { Todo, TodoStatus } from "models/todo";
import { EnhanceTodoStatus } from "pages/ToDoPage";
import { isTodoCompleted } from "utils";
import {
  Button,
  ButtonColors,
  Checkbox,
  ModalConfirm,
  Paper,
} from "components/commons";
import cn from "classnames";
import "./style.css";

interface Props {
  todos: Todo[];
  showing: EnhanceTodoStatus;
  onFilter: (status: EnhanceTodoStatus) => void;
  onDeleteAllToDo: () => void;
  onToggleAllToDo: (checked: boolean) => void;
}

const TAB_FILTER = [
  {
    name: "All",
    value: "ALL",
  },
  {
    name: "Active",
    value: TodoStatus.ACTIVE,
  },
  {
    name: "Completed",
    value: TodoStatus.COMPLETED,
  },
];

const ToolbarToDo = (props: Props) => {
  const { showing, todos, onFilter, onDeleteAllToDo, onToggleAllToDo } = props;
  const [showModalConfirm, setShowModalConfirm] = useState(false);

  const activeTodos = todos.reduce(function (accum, todo) {
    return isTodoCompleted(todo) ? accum : accum + 1;
  }, 0);

  function _toggleModalConfirm() {
    setShowModalConfirm((prev) => !prev);
  }

  return (
    <Paper className="Todo__toolbar p10">
      {todos.length ? (
        <div className="Action__checkAll">
          <Checkbox
            checked={activeTodos === 0}
            onChange={(e) => onToggleAllToDo(e.target.checked)}
          />
        </div>
      ) : (
        <div />
      )}
      <div className="Todo__tabs">
        {TAB_FILTER.map((tab, index) => (
          <div
            key={index}
            className={cn(
              "Action__btn",
              showing === tab.value && "Action__btn-selected"
            )}
            onClick={() => onFilter(tab.value as EnhanceTodoStatus)}
          >
            {tab.name}
          </div>
        ))}
      </div>
      <Button
        disabled={!todos.length}
        color={ButtonColors.danger}
        onClick={_toggleModalConfirm}
      >
        Clear all
      </Button>

      <ModalConfirm
        onConfirm={() => {
          onDeleteAllToDo();
          _toggleModalConfirm();
        }}
        onClose={_toggleModalConfirm}
        show={showModalConfirm}
      />
    </Paper>
  );
};

export default ToolbarToDo;
