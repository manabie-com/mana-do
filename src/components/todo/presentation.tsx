/**
 * For applying presentation component pattern
 * Split UI process and logic process.
 */
import React from 'react';
import TodoItem from './subPresentations/todoItem';
import { Todo } from 'root/models/todo';

export type TodoProps = {
  inputRef: any,
  onCreateTodo: React.KeyboardEventHandler,
  showTodos: Array<Todo>,
  onUpdateTodoStatus: Function,
  deleteTodo: Function,
  todos: Array<Todo>,
  activeTodos: Number,
  onToggleAllTodo: React.ChangeEventHandler,
  handleShowingAll: React.MouseEventHandler,
  handleShowingActive: React.MouseEventHandler,
  handleShowingCompleted: React.MouseEventHandler,
  onDeleteAllTodo: React.MouseEventHandler,
}

const text = {
  /** help to concentrate all text, easy to edit text or apply multi-language feature */
  x: 'X',
  all: 'All',
  active: 'Active',
  completed: 'Completed',
  clearAllTodos: 'Clear all todos',
  whatNeedToBeDone: 'What need to be done?'
}


const TodoPresentation = (props: TodoProps) => {
  const {
    inputRef, onCreateTodo, showTodos
    , onUpdateTodoStatus, deleteTodo, todos
    , activeTodos, onToggleAllTodo
    , handleShowingAll, handleShowingActive
    , handleShowingCompleted, onDeleteAllTodo
  } = props

  return (
    <div className='ToDo__container'>
      <div className='Todo__creation'>
        <input
          ref={inputRef}
          className='Todo__input'
          placeholder={text.whatNeedToBeDone}
          onKeyDown={onCreateTodo}
        />
      </div>
      <div className='ToDo__list'>
        {
          showTodos.map((todo: Todo, index: number) => (
            <TodoItem key={`show-todo-item-${index}`}
              todo={todo}
              deleteItemText={text.x}
              deleteTodo={deleteTodo}
              onUpdateTodoStatus={onUpdateTodoStatus}
            />
          ))
        }
      </div>
      <div className='Todo__toolbar'>
        {todos.length > 0 ?
          <input
            type='checkbox'
            checked={activeTodos === 0}
            onChange={onToggleAllTodo}
          /> : <div />
        }
        <div className='Todo__tabs'>
          <button className='Action__btn' onClick={handleShowingAll}>
            {text.all}
          </button>
          <button className='Action__btn' onClick={handleShowingActive}>
            {text.active}
          </button>
          <button className='Action__btn' onClick={handleShowingCompleted}>
            {text.completed}
          </button>
        </div>
        <button className='Action__btn' onClick={onDeleteAllTodo}>
          {text.clearAllTodos}
        </button>
      </div>
    </div>
  );
};

//use shallowly compare complex objects in the props object
export default React.memo(TodoPresentation)