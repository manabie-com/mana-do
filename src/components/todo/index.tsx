import React, { useEffect, useReducer, useRef, useState, useMemo, useCallback } from 'react';
import { RouteComponentProps } from 'react-router-dom';

import reducer, { initialState } from 'root/store/reducer';
import {
  setTodos,
  createTodo,
  deleteTodo,
  toggleAllTodos,
  deleteAllTodos,
  updateTodoStatus,
  updateTodoContent
} from 'root/store/actions';
import Service from 'root/service';
import { TodoStatus } from 'root/models/todo';
import { isTodoCompleted } from 'root/utils';
import TodoPresentation from './presentation';

import './_todo.css'

type EnhanceTodoStatus = TodoStatus | 'ALL';


const TodoComponent = ({ history }: RouteComponentProps) => {
  const [{ todos }, dispatch] = useReducer(reducer, initialState);
  const [showing, setShowing] = useState<EnhanceTodoStatus>('ALL');
  const inputRef = useRef<HTMLInputElement>(null);

  useEffect(() => {
    (async () => {
      const resp = await Service.getTodos();
      dispatch(setTodos(resp || []));
    })()
  }, [])

  useEffect(() => {
    if (todos.length > 0) {
      const lastItem = document.getElementById(`todo-item-${todos[todos.length - 1].id}`)
      if (lastItem) setTimeout(() => {
        lastItem.scrollIntoView({ block: 'end',  behavior: 'smooth' })
      }, 401) // delay to wait for css animation work first
    }
  }, [todos, showing])

  const onCreateTodo = useCallback(async (e: React.KeyboardEvent<HTMLInputElement>) => {
    /**
     * reduce rebinding function. Does not rebind when dependencies has not changed
     */
    if (e.key === 'Enter' && inputRef.current) {
      try {
        const resp = await Service.createTodo(inputRef.current.value);
        dispatch(createTodo(resp));
        inputRef.current.value = '';
      } catch (e) {
        if (e.response.status === 401) {
          history.push('/')
        }
      }
    }
  }, [dispatch, history])

  const onUpdateTodoStatus: Function = useCallback(async (e: React.ChangeEvent<HTMLInputElement>, todoId: string) => {
    try {
      await Service.updateTodoStatus(todoId, e.target.checked);
      dispatch(updateTodoStatus(todoId, e.target.checked));
    } catch (error) {

    }
  }, [dispatch])

  const onToggleAllTodo = useCallback(async (e: React.ChangeEvent<HTMLInputElement>) => {
    /**
     * reduce rebinding function. Does not rebind when dependencies has not changed
     */
    try {
      await Service.toggleAllTodos(e.target.checked)
      dispatch(toggleAllTodos(e.target.checked))
    } catch (error) {

    }
  }, [dispatch])

  const onDeleteAllTodo = useCallback(async () => {
    /**
     * reduce rebinding function. Does not rebind when dependencies has not changed
     */
    try {
      await Service.deleteAllTodo()
      dispatch(deleteAllTodos());
    } catch (error) {

    }
  }, [dispatch])

  const handleDeleteTodo = useCallback(async (todoId: string) => {
    /**
     * reduce rebinding function. Does not rebind when dependencies has not changed
     */
    try {
      await Service.deleteTodo(todoId)
      dispatch(deleteTodo(todoId));
    } catch (error) {

    }
  }, [dispatch])

  const handleUpdateTodoContent = useCallback(async (todoId: string, content: string) => {
    /**
     * reduce rebinding function. Does not rebind when dependencies has not changed
     */
    try {
      await Service.updateTodoContent(todoId, content)
      dispatch(updateTodoContent(todoId, content));
    } catch (error) {

    }
  }, [dispatch])

  const handleShowingAll = useCallback(() => {
    /**
     * reduce rebinding function. Does not rebind when dependencies has not changed
     */
    setShowing('ALL')
  }, [setShowing])

  const handleShowingActive = useCallback(() => {
    /**
     * reduce rebinding function. Does not rebind when dependencies has not changed
     */
    setShowing(TodoStatus.ACTIVE)
  }, [setShowing])

  const handleShowingCompleted = useCallback(() => {
    /**
     * reduce rebinding function. Does not rebind when dependencies has not changed
     */
    setShowing(TodoStatus.COMPLETED)
  }, [setShowing])

  const showTodos = useMemo(() => todos.filter((todo) => {
    /**
     * Reduce recalculating showing todos, Does not recalculate showing todos when todos and showing state have not changed
     */
    switch (showing) {
      case TodoStatus.ACTIVE:
        return todo.status === TodoStatus.ACTIVE;
      case TodoStatus.COMPLETED:
        return todo.status === TodoStatus.COMPLETED;
      case 'ALL':
      default:
        return true;
    }
  }), [showing, todos])

  const activeTodos = useMemo(() => todos.reduce(function (accum, todo) {
    /**
     * Reduce recalculating activeTodos, Does not recalculate active todos when todos has not changed
    */
    return isTodoCompleted(todo) ? accum : accum + 1;
  }, 0), [todos])

  return (
    <TodoPresentation
      inputRef={inputRef}
      onCreateTodo={onCreateTodo}
      showTodos={showTodos}
      onUpdateTodoStatus={onUpdateTodoStatus}
      deleteTodo={handleDeleteTodo}
      todos={todos}
      activeTodos={activeTodos}
      onToggleAllTodo={onToggleAllTodo}
      handleShowingAll={handleShowingAll}
      handleShowingActive={handleShowingActive}
      handleShowingCompleted={handleShowingCompleted}
      onDeleteAllTodo={onDeleteAllTodo}
      onUpdateTodoContent={handleUpdateTodoContent}
      showing={showing}
    />
  );
};

export default TodoComponent;