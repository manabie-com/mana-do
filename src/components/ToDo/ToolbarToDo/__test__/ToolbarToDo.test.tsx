import React from "react";
import ReactDOM from "react-dom";
import { render, cleanup } from "@testing-library/react";

import ToolbarToDo from "components/ToDo/ToolbarToDo";
import userEvent from "@testing-library/user-event";
import { TodoStatus } from "models/todo";
import { EnhanceTodoStatus } from "pages/ToDoPage";
import renderer from "react-test-renderer";
import "@testing-library/jest-dom";

const propsTest = {
  todos: [
    {
      id: "todo",
      user_id: "user_id",
      content: "content",
      status: TodoStatus.ACTIVE,
      created_date: new Date().toISOString(),
    },
  ],
  onFilter: (status: EnhanceTodoStatus) => null,
  showing: TodoStatus.ACTIVE,
  onDeleteAllToDo: jest.fn(),
  onToggleAllToDo: jest.fn(),
};

let container: HTMLElement;

beforeEach(() => {
  container = render(<ToolbarToDo {...propsTest} />).container;
});

afterEach(cleanup);

test("ToolbarToDo: render without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<ToolbarToDo {...propsTest} />, div);
  ReactDOM.unmountComponentAtNode(div);
});

test("ToolbarToDo: test btn clear all", () => {
  const btnClearAllEl = container.querySelector("button") as HTMLButtonElement;
  expect(btnClearAllEl).toHaveTextContent("Clear all");

  userEvent.click(btnClearAllEl);
  const modalEl = container.querySelector(".modal-custom") as HTMLButtonElement;
  expect(modalEl).toHaveClass("display-block");
});

test("ToolbarToDo: match snapshot", () => {
  const tree = renderer.create(<ToolbarToDo {...propsTest} />).toJSON();
  expect(tree).toMatchSnapshot();
});
