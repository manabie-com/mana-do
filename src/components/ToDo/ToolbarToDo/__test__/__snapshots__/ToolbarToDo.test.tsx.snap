// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolbarToDo: match snapshot 1`] = `
<div
  className="paper-custom Todo__toolbar p10 paper-custom-elevation-1"
>
  <div
    className="Action__checkAll"
  >
    <label
      className="checkbox-custom"
      data-testid="checkbox-custom"
    >
      <svg
        className="icon_GG7Y icon_3JOC"
        height="24"
        version="1.1"
        viewBox="0 0 32 32"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      >
        <path
          d="M26.667 28.444H5.334c-.978 0-1.778-.8-1.778-1.778V5.333c0-.978.8-1.778 1.778-1.778h21.333c.978 0 1.778.8 1.778 1.778v21.333c0 .978-.8 1.778-1.778 1.778zM28.444 0H3.555A3.566 3.566 0 00-.001 3.556v24.889c0 1.956 1.6 3.556 3.556 3.556h24.889c1.956 0 3.556-1.6 3.556-3.556V3.556C32 1.6 30.4 0 28.444 0z"
          fill="#1caba0"
        />
      </svg>
      <div
        className="root-input checkbox-custom-input"
      >
        <input
          checked={false}
          className="input-custom input-size-md"
          onChange={[Function]}
          readOnly={true}
          type="checkbox"
        />
      </div>
    </label>
  </div>
  <div
    className="Todo__tabs"
  >
    <div
      className="Action__btn"
      onClick={[Function]}
    >
      All
    </div>
    <div
      className="Action__btn Action__btn-selected"
      onClick={[Function]}
    >
      Active
    </div>
    <div
      className="Action__btn"
      onClick={[Function]}
    >
      Completed
    </div>
  </div>
  <button
    className="btn-custom btn-custom-danger"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Clear all
  </button>
  <div
    className="modal-custom display-none"
  >
    <section
      className="modal-main"
      style={
        Object {
          "width": "30%",
        }
      }
    >
      <div
        className="header-modal mb40"
      >
        <span
          className="title"
        >
          Are you sure delete ?
        </span>
        <span
          className="icon-close"
          onClick={[Function]}
        >
          <img
            alt="icon-close"
            src="images/cancel.png"
            width="30"
          />
        </span>
      </div>
      <div
        className="content-modal mt30"
      >
        <div
          className="content-modal-confirm"
        >
          Item will be permanently deleted?
        </div>
        <div
          className="action-modal-confirm"
        >
          <button
            className="mr10 btn-custom btn-custom-danger"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Yes
          </button>
          <button
            className="btn-custom btn-custom-primary"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            No
          </button>
        </div>
      </div>
    </section>
  </div>
</div>
`;
